src/main/java/reactor/kafka/receiver/observation/ReceiverObservations.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2016-2023 VMware Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.kafka.receiver.observation;
+
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.ObservationRegistry;
+import io.micrometer.observation.contextpropagation.ObservationThreadLocalAccessor;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import reactor.core.publisher.Mono;
+import reactor.util.annotation.Nullable;
+
+/**
+ * Helper factories for {@link ConsumerRecord} observation in the end-user code.
+ * Typically used like this on a receiver {@link reactor.core.publisher.Flux}
+ * result in a {@link Mono#transformDeferred(Function)} operator for each record:
+ * <pre>
+ * {@code
+ * Flux<ReceiverRecord<Integer, String>> receive =
+ *        KafkaReceiver.create(receiverOptions.subscription(Collections.singletonList(topic)))
+ *                .receive()
+ *                .flatMap(record ->
+ *                        Mono.just(record)
+ *                                .transformDeferred(mono ->
+ *                                        ReceiverObservations.observe(mono, record, "reactor kafka receiver",
+ *                                                bootstrapServers(), OBSERVATION_REGISTRY)));
+ * }
+ * </pre>
+ * <p>
+ * These factories are needed if there is a requirements to process a record in a reactive manner within
+ * the mentioned consumer observation.
+ * Otherwise, a regular {@link KafkaReceiverObservation} API is enough to use
+ * in a {@link reactor.core.publisher.Flux#doOnNext(Consumer)} operator.
+ *
+ * @see KafkaReceiverObservation
+ */
+public final class ReceiverObservations {
+
+    public static <T> Mono<T> observe(Mono<T> userMono, ConsumerRecord<?, ?> record, String clientId,
```

Comment:
Could we make this into a Function<Flux<T>, Flux<T>> return so that it's easier to use with transformDeferred ? or maybe some tapping like it's in reactor-core ? I believe currently it's recommended to use Flux...tap(Micrometer.observation(observationRegistry)), wouldn't it be nice to make use of same "technique" ?

Reply:
Well, I probably don't understand `tap()` at all. So, would be great if you share more info and your thoughts.
My idea with this function is to let end-user to observe every `ConsumerRecord` withing its own `Mono` and an `Observation` tied to its context. How is the could be possible with the whole `Flux`?
How `tap()` can help us here, please?

Reply:
I'm not that proficient with it yet, but i've seen reactor team using it. If i understand it properly, they are using signallistener registered with tap to start observations on doFirst, doOnNext and closing them based on complete signal (doOnCancel, doOnComplete, doOnError etc.). But I didn't have time to properly analyse it all.

Here's example of their implementation:

https://github.com/reactor/reactor-core/blob/main/reactor-core-micrometer/src/main/java/reactor/core/observability/micrometer/MicrometerMeterListenerConfiguration.java

https://github.com/reactor/reactor-core/blob/main/reactor-core-micrometer/src/main/java/reactor/core/observability/micrometer/MicrometerObservationListener.java

Reply:
@maciej-gromul 

> Could we make this into a Function<Flux, Flux> return so that it's easier to use with transformDeferred ? or maybe some tapping like it's in reactor-core ? I believe currently it's recommended to use Flux...tap(Micrometer.observation(observationRegistry)), wouldn't it be nice to make use of same "technique" ?

Have you seen this post? It does exactly what you are mentioning:

https://github.com/reactor/reactor-kafka/pull/325#issuecomment-1460517792

Reply:
@Tomasz-Marciniak I've seen that, I'm not that good with kotlin but I think this is similar to what i've written. Generally this is how it should look like if someone want's to instrument the tracing manually to the receivers. 

```java
receiver.receive()
  .tap(SomeObservationListener.observation())

//or

receiver.receive()
  .transformDeferred(tracingContextDeferMethodFromSomewhere)
```

So lessening the possible boilerplate for api clients. Maybe we could even add it to KafkaReceiver implementation as  part of ReceiverOptions ? That way someone would simply do:

```java
KafkaReceiver.create(ReceiverOptions.create().observable(SomeObservableConfiguration))
```

and then under the hood DefaultKafkaReceiver can simply check that configuration and if needed apply tracing transformation. If someone wants to have more control they can simply not register that and manually execute that tap/transformDeferred

Reply:
Thank you for feedback!

One question though: does it work per consumer record?
The idea is to be able to restore a trace from produce and apply it for just consumed record.
I'm not sure that in most case it would be just enough to record a span for consumed record and give up.
Probably end-user would like to continue a processing with that observation on a record and produce more spans downstream.

Reply:
There's this implementation on reactor-core I find to be really nice [MicrometerObservationListener](https://github.com/reactor/reactor-core/blob/main/reactor-core-micrometer/src/main/java/reactor/core/observability/micrometer/MicrometerObservationListener.java) that is simply used as 
```java
Flux.just(1,2,3)
  .name("something")
  .tag("foo", "bar")
  .tap(Micrometer.observation(registry))
```

Needs the least knowledge from api users and looks rather clean :P

Reply:
> One question though: does it work per consumer record?

I'm using transformDeferred with sleuth reactor on spring boot 2, and it works with brave instrumentation. I'm not sure about tap in spring boot 3 + micrometer. I didn't get to test it because for me current observations context doesn't work properly even with standard publishers, for some reason spans end up not propagating the context correctly downstream.

Reply:
OK. Sorry for delay: was busy with other stuff.
So, I read a bit about `tap()` and saw how it works in the source code.
My conclusion that it has very minimal functionality for our use-case.
And it is here like this: the `ConsumerRecord` has an info about a trace. 
It would be great to be able to restore that trace in the consumer and start a new child span.

The `tap()` wants from us to do something on subscription to the `Publisher`. 
For example, the mentioned `Micrometer.observation()` does start an `Observation` at that point.
All the upstream operators in the `Flux/Mono` will be aware of that observation then.
This observation is stopped in the `doOnNext` for `Mono` and in the `doOnComplete` for `Flux`.
Therefore everything after the `tap()` is not aware about that observation already.

If you look into my `ConsumerRecord` observation requirement, you'll see that this `tap()` won't fit.
Sure, we probably can implement something like "just restore and record" hook, then a simple `doOnNext` for that `tap` listener would be enough.
But is it a goal we would try to reach here? The observation will be restored from the record, started and stopped immediately. Just because this is a nature how the `tap()` works.
It is probably going to be fully similar what we have with a `handle()`. Therefore no reason to abuse `tap()` where we won't get any gains from its purpose.

So, what do I miss with the `tap()`, please?


Sounds like `transformDeferred` is still only the way for us to be able to restore such an observation from a consumer record and let it to be spread around some end-users operator. 
We might be need to think about eliminating an extra `ConsumerRecord` argument for that function and something like `Flux` input style would be great as well.
But this is something what remains to be explored.

Thank you for understanding!

Reply:
I've had some issues created for observations with tap and like you say it works only in a context of whole publisher while propagating the context only downstream. I've created suggestion/request issue to give some kind of ability to publish the context upstream like it was done with spring sleuth but right now it's impossible. Current usage of observations in case of flux that should behave like a mono requires either having the main contract as Flux<Mono<T>>receive(), then each mono will have it's own context that consumers can read in their flatmaps or having callable for either flatmap / transform to execute, both changing the standard way reactor pipelines are written, forcing unintuitive schemas in case we ever need telemetry.

Reply:
Potentially we could write custom implementation `KafkaSubscriber implements InnerOperator<T, T>` that could for each element read from it's internal source change the context available in methods onNext/tryOnNext and handle errors with onError/cancel/onComplete. But that would need contribution on ReactorCore since InnerOperator is package scoped.

Reply:
Fingers crossed for this